BookingsApi/BookingsApi.DAL/Services/HearingAllocationService.cs

```diff
@@ -318,5 +325,74 @@ private JusticeUser SelectRandomly(IList<JusticeUser> csos)
             var cso = csos[csoIndex-1];
             return cso;
         }
+        
+        private static Dictionary<string, int> BuildUsernameConcurrencyCountDictionary(List<VideoHearing> hearings)
+        {
+            // builds a dictionary of usernames and a boolean indicating whether the user has 
+            // concurrent hearings exceeding the 'concurrencyLimit'
+            
+            // get a distinct list of all users for the current collection of hearings
+            var vhoUsernames = hearings.Where(x => x.AllocatedTo != null).Select(x => x.AllocatedTo.Username).Distinct();
+            
+            var vhoUsernamesConcurrencyExceeded = new Dictionary<string, int>();
+
+            // for each username, determine if, for the current collection of hearings, it has more than 3 overlapping hearings
+            foreach (var username in vhoUsernames)
+            {
+                // get all the hearings allocated to the username
+                var hearingsForVho = hearings.Where(x => x.AllocatedTo?.Username == username)
+                    .OrderBy(hearing => hearing.ScheduledDateTime).ToList();
+                
+                // a counter to count overlapping hearings
+                var numberOfOverlappingHearings = 0;
+                
+                // there can't be an overlap if there's only 1 hearing, return false for this username
+                if (hearingsForVho.Count <= 1)
+                {
+                    vhoUsernamesConcurrencyExceeded.Add(username, 0);
+                    return vhoUsernamesConcurrencyExceeded;
+                }
+                
+                for (var i = 0; i < hearingsForVho.Count; i++)
```

Comment:
could maybe move this for loop into a separate method to condense this a bit. Might fix the sonar cloud warning:

`numberOfOverlappingHearings = CountOverlappingHearings(hearingsForVho);`